onadata/libs/utils/logger_tools.py

```diff
@@ -240,6 +240,13 @@ def save_attachments(xform, instance, media_files):
                 mimetype=content_type,
                 name=filename,
                 extension=extension)
+        replaced_attachments = (
+            a for a in Attachment.objects.filter(instance=instance)
+            if a.media_file.name.split('/')[-1] not in media_files)
```

Comment:
I believe there's a way to push this all into a single query, but it may require a raw SQL query (it may not, so worth investigating that first) and it may not be worthwhile, since I'm not sure about the performance of the SQL operators used, with those warnings in place..

It is generally faster to use a single query instead of a loop. You can put the string search into a where clause searching a joined `media_file` table with something like

```sql
REVERSE(SPLIT_PART(REVERSE(a.media_file.name, '/', 1))) NOT LIKE %{media_files.join('|')}
```

that's pseudocode, but hopefully the idea is clear, the inner reverse to find the last occurs of the delimiter, the outer reverse so it's in the correct other for the LIKE clause that joins all the media file text.

Reply:
After some research, I am thinking of working with django's [F expression](https://docs.djangoproject.com/en/3.0/ref/models/expressions/#f-expressions) for this. I mean, we could still  use raw sql, but i have a feeling the queryset api could adequately handle this, for this use case.

Using F expressions allows you to perform manipulations on model fields without loading them into memory.
So a clean way to filter and update this attachments could be:

`Attachment.objects.filter(instance=instance).filter(~Q(name_in=filenames).update(deleted_at=F('deleted_at') timezone.now())`

Here we could filter out the queryset for:
1. Attachments for that specific submission
2. Attachment that are not present in the media files recieved

and consequently update the deleted field for those attachments, in a way 'soft deleting' those attachments.

But I guess this could work too:
`Attachment.objects.filter(instance=instance).filter(~Q(name_in=filenames).update(deleted_at=timezone.now())`

without adding another layer of complexity to the query.
Any take on this?

Reply:
This looks great, I'm pretty sure Django will lazy-evaluate this query, which means you can split this into 2 steps

1) define the queryset
2) pass that queryset to a soft_delete function

so in this code you'd define the queryset and pass it:

```python
to_delete_queryset = Attachment.objects.filter(instance=instance).filter(~Q(name_in=filenames)
Attachment.soft_delete()
```

or actually, in a single call is probably fine:

```python
Attachment.soft_delete(Attachment.objects.filter(instance=instance).filter(~Q(name_in=filenames))
```



then in `models/attachment.py` add the class function

```python
@classmethod
def safe_delete(cls, queryset)
    queryset.update(deleted_at=timezone.now())
```

I think you can overload class/non-class methods like that, not sure if it's best practice.

The reason I prefer splitting it up, is so that if we can modify the logic for soft deletion in one place in the future, instead of having to comb through the code and find place where we set `deleted_at`, e.g. suppose we change the name of the `deleted_at` column, or more likely suppose we want to explicitly log or call a metric service when we soft delete something.